### ISSUE TITLE
error: 'dirs' undeclared on Solaris in lmfile module on rsyslog 8.11.0

### ISSUE THREAD

**Issue by dago** [rsyslog/rsyslog#417]
Enabling `--enable-imsolaris` and `--enable-imfile` on Solaris 10 Sparc leads to the following compilation error in syslog 8.11.0:

```
gmake[2]: Entering directory '/home/dam/mgar/pkg/rsyslog/trunk/work/solaris10-sparc/build-isa-sparcv8plus/rsyslog-8.11.0/plugins/imsolaris'
  CC       imsolaris_la-imsolaris.lo
imsolaris.c: In function 'readLog':
imsolaris.c:214:3: warning: 'submitMsg' is deprecated (declared at ../../dirty.h:32) [-Wdeprecated-declarations]
   CHKiRet(submitMsg(pMsg));
   ^
In file included from imsolaris.c:78:0:
imsolaris.c: At top level:
../../runtime/module-template.h:114:17: warning: 'modGetCnfName' defined but not used [-Wunused-function]
 static rsRetVal modGetCnfName(uchar **cnfName) \
                 ^
imsolaris.c:89:1: note: in expansion of macro 'MODULE_CNFNAME'
 MODULE_CNFNAME("imsolaris")
 ^
../../runtime/module-template.h:851:17: warning: 'beginCnfLoad' defined but not used [-Wunused-function]
 static rsRetVal beginCnfLoad(modConfData_t **ptr, __attribute__((unused)) rsconf_t *pConf)\
                 ^
imsolaris.c:309:1: note: in expansion of macro 'BEGINbeginCnfLoad'
 BEGINbeginCnfLoad
 ^
../../runtime/module-template.h:899:17: warning: 'endCnfLoad' defined but not used [-Wunused-function]
 static rsRetVal endCnfLoad(modConfData_t *ptr)\
                 ^
imsolaris.c:314:1: note: in expansion of macro 'BEGINendCnfLoad'
 BEGINendCnfLoad
 ^
../../runtime/module-template.h:921:17: warning: 'checkCnf' defined but not used [-Wunused-function]
 static rsRetVal checkCnf(modConfData_t *ptr)\
                 ^
imsolaris.c:319:1: note: in expansion of macro 'BEGINcheckCnf'
 BEGINcheckCnf
 ^
../../runtime/module-template.h:959:17: warning: 'activateCnf' defined but not used [-Wunused-function]
 static rsRetVal activateCnf(modConfData_t *ptr)\
                 ^
imsolaris.c:324:1: note: in expansion of macro 'BEGINactivateCnf'
 BEGINactivateCnf
 ^
../../runtime/module-template.h:977:17: warning: 'freeCnf' defined but not used [-Wunused-function]
 static rsRetVal freeCnf(void *ptr)\
                 ^
imsolaris.c:329:1: note: in expansion of macro 'BEGINfreeCnf'
 BEGINfreeCnf
 ^
  CC       imsolaris_la-sun_cddl.lo
  CCLD     imsolaris.la
gmake[2]: Leaving directory '/home/dam/mgar/pkg/rsyslog/trunk/work/solaris10-sparc/build-isa-sparcv8plus/rsyslog-8.11.0/plugins/imsolaris'
Making all in plugins/imfile
gmake[2]: Entering directory '/home/dam/mgar/pkg/rsyslog/trunk/work/solaris10-sparc/build-isa-sparcv8plus/rsyslog-8.11.0/plugins/imfile'
  CC       imfile_la-imfile.lo
imfile.c: In function 'modExit':
imfile.c:1872:5: error: 'dirs' undeclared (first use in this function)
  if(dirs != NULL) {
     ^
imfile.c:1872:5: note: each undeclared identifier is reported only once for each function it appears in
imfile.c:1877:7: error: 'wdmap' undeclared (first use in this function)
  free(wdmap);
       ^
imfile.c: At top level:
imfile.c:824:1: warning: 'lstnDup' defined but not used [-Wunused-function]
 lstnDup(lstn_t ** ppExisting, uchar *const __restrict__ newname)
 ^
Makefile:498: recipe for target 'imfile_la-imfile.lo' failed
gmake[2]: *** [imfile_la-imfile.lo] Error 1
```

**Comment by h0lzi** [comment]
this patch might help:

https://github.com/joyent/pkgsrc/blob/trunk/sysutils/rsyslog/patches/patch-plugins_imfile_imfile.c

This is form here: https://github.com/rsyslog/rsyslog/issues/352

**Comment by dago** [comment]
Good catch, that did the trick. Should be merged upstream IMHO.

**Comment by alorbach** [comment]
We had the same problem on Freebsd while testing, it seems to be the same as here: 
So the problem should be fixed with this commit https://github.com/rsyslog/rsyslog/commit/2edfb087b848af1bc2df1369807a1e6413789163

Can somebody verify/test this against Solaris?

**Comment by dago** [comment]
The patch from joyent does a little more as setting `iRet = doPolling();` in the else-case.

**Comment by rgerhards** [comment]
That looks like it either was for an old version or is wrong. If you look at the code, there is a stub for do_inotify(): https://github.com/rsyslog/rsyslog/blob/6cb712d044c02044668dabce810019a8d74ba20f/plugins/imfile/imfile.c#L1747

**Comment by rgerhards** [comment]
I should add that we cannot blindly use polling mode without telling the users. Polling mode does not offer all features, so users need to know what is going on.

**Comment by rgerhards** [comment]
fixed in 8.12.0.
